src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -11,44 +11,38 @@ private unsafe string GetAsciiCore(char* unicode, int count)
             uint flags = Flags;
             CheckInvalidIdnCharacters(unicode, count, flags, nameof(unicode));
 
-            const int StackAllocThreshold = 512;
-            if (count < StackAllocThreshold)
+            // Each unicode character is represented by 2 ASCII chars
+            // and a "xn--" prefix of length 4
+            int estimatedLength = count * 2 + 4;
+            int actualLength;
+            if (estimatedLength < StackallocThreshold)
             {
-                char* output = stackalloc char[count];
-                return GetAsciiCore(unicode, count, flags, output, count, reattempt: true);
-            }
-            else
-            {
-                char[] output = new char[count];
-                fixed (char* pOutput = output)
+                char* outputSmall = stackalloc char[estimatedLength];
+                actualLength = Interop.GlobalizationNative.ToAscii(flags, unicode, count, outputSmall, estimatedLength);
+                if (actualLength == 0)
                 {
-                    return GetAsciiCore(unicode, count, flags, pOutput, count, reattempt: true);
+                    throw new ArgumentException(SR.Argument_IdnIllegalName, nameof(unicode));
+                }
+                else if (actualLength <= estimatedLength)
+                {
+                    return new string(outputSmall, 0, actualLength);
                 }
             }
-        }
-
-        private unsafe string GetAsciiCore(char* unicode, int count, uint flags, char* output, int outputLength, bool reattempt)
-        {
-            int realLen = Interop.GlobalizationNative.ToAscii(flags, unicode, count, output, outputLength);
-
-            if (realLen == 0)
-            {
-                throw new ArgumentException(SR.Argument_IdnIllegalName, nameof(unicode));
-            }
-            else if (realLen <= outputLength)
+            else
             {
-                return new string(output, 0, realLen);
+                actualLength = Interop.GlobalizationNative.ToAscii(flags, unicode, count, null, 0);
```

Comment:
I missed this in my pseudo-code, but we need the comparison-against-0-and-throw here as well.  If you wanted, you could remove the previous check, replacing the return with ```if (actualLength > 0 && actualLength <= estimatedLength)```, and then have the ```if (actualLength <= 0) throw ...``` be after this else block, so that it applies to both cases.